BackPropagation test with one hidden layer.

cashecollins opened the pull request:
Back propagation test with one hidden layer and various training inputs, I wasn't sure how to get our information nor was I sure about the structure of our network so I just have 1 hidden layer, but we can modify it if we want to have more.

cashecollins committed:
BackPropagation test with one hidden layer.

Back propagation test with one hidden layer and various training inputs.

ktrinanm commented:
I would remove the .class file from the commit.. i also suggest making sure this class can be easily integrated with the other code we have so far.

cashecollins committed:
Easy Integration

I made the backpropagation more easily integratable to the groups code.

cashecollins commented:
So I know you are all asleep at this point, so we can talk about it tomorrow, but I have a couple questions about the structure/number of hidden layers that we are using. As well as how we are accessing those nodes.

ktrinanm commented:
Hey I just looked at this, but I'm going to wait until you add any changes you made today before I review it. Thanks!

ktrinanm commented:
Hey where are you at on finishing the integration changes?

cashecollins committed:
Integration

Integrations

cashecollins committed:
Changed Class Name

cashecollins committed:
Updated Back propagation code

I updated the backpropagate method, however I've not been able to get output that makes sense.

ktrinanm commented on backpropagation.java:
```diff
@@ -0,0 +1,438 @@
+        int[] r = new int[2]; //Target is the expected output neuron for each data set.
+        z = neurons[1];
+
+        if (classification[n] == 1){
```
So, here, you're reading from the classification array and not from the arrayList that has the training data. That could be causing problems.

cashecollins replied:
Thanks! I made the change.

ktrinanm replied:
Cool. Let me know if it helps. did talking to tadayon help at all?

cashecollins replied:
I was really busy and the free time that I had, he was busy so I wasn't able to talk to him

cashecollins replied:
I've been debugging though and I think that the weights are updating correctly. except that like joe was saying i think we have doubles overflowing and rounding to 1.0 or 0.0 and that might be messing with the code, I tried changing the sigmoid scale and that helped a little but still there was some rounding.

ktrinanm replied:
So how close are we to being able to get results from it?

ktrinanm commented on backpropagation.java:
```diff
@@ -0,0 +1,438 @@
+
+    //Linear Algebra Helper Functions
+    private static double vectorTransposeMultiplication(double[] A, double[] B){
+        double result = 0.0;
```
Before completing this, I would verify that A.length == B.length because it could bring an issue to light if that's false.

cashecollins replied:
Alright I looked at this and made some changes so that the weights update to the correct spots.

ktrinanm reviewed (commented):
Maybe checkout these comments to see if the mentioned lines could be causing issues?

cashecollins commented on backpropagation.java:
```diff
@@ -0,0 +1,438 @@
+
+        double sum = 0.0;
+        for (int h = 0; h < INPUT_NEURONS; h++){
+            double s = vectorTransposeMultiplication(weights[1][h], x);
```
nah, the way that our weights are initialized  wweights[0] is an array full of 0.0, weights[1] are the weights from input to hidden, and weights[2] are the weights from hidden to output.

ktrinanm replied:
Interesting. That might be something we need to look into updating

cashecollins committed:
Updated backpropagation to mirror psuedo code given in class

ktrinanm committed:
Changing to 1 output

ktrinanm committed:
Working through logic and changing to 1 output class

ktrinanm committed:
Tried changing to use sigmoid where necessary

cashecollins committed:
Working(ish) single hidden layer Neural Network

Should be the final edit for method 1, with 1 hidden layer.

cashecollins committed:
Merge branch 'V1.backpropagation' of https://github.com/ktrinanm/ML_NeuralNetwork into V1.backpropagation

Idk what i'm doing

cashecollins committed:
Code Cleanup

Cleaned up the code, got rid of a bunch of print statements, made coments to make code more understandable
